src/Week1/HW01.hs

```diff
@@ -14,38 +13,99 @@ dropLastDigit n = n `div` 10
 
 -- Exercise 2 -----------------------------------------
 
+{- toRevDigits should convert positive Integers to a list of digits. (For
+0 or negative inputs, toRevDigits should return the empty list.) -}
+
+-- non tail recursive
 toRevDigits :: Integer -> [Integer]
-toRevDigits n = helper n []
+toRevDigits n 
+        | n <= 0 = [] 
+        | otherwise = lastDigit n : toRevDigits (dropLastDigit n)
+
+-- tail recursive version, assuming reverse is tail recursive
+toRevDigits' :: Integer -> [Integer]
+toRevDigits' n = helper n []
     where
-        helper x acc 
-            | x <= 0 = reverse acc 
+        helper :: Integer -> [Integer] -> [Integer]
+        helper x acc
+            | x <= 0 = reverse acc
+            | otherwise = helper (dropLastDigit x) (lastDigit x : acc)
+
+-- tail recursive non reversed digits
+toDigits :: Integer -> [Integer]
+toDigits n = helper n []
+    where
+        helper :: Integer -> [Integer] -> [Integer]
+        helper x acc
+            | x <= 0 = acc
             | otherwise = helper (dropLastDigit x) (lastDigit x : acc)
     
 
 -- Exercise 3 -----------------------------------------
 
 -- Double every second number in a list starting on the left.
+-- Original solution, note to self, avoid using head and tail in the future, can cause runtime exceptions
 doubleEveryOther :: [Integer] -> [Integer]
 doubleEveryOther lst = helper 0 lst [] 
     where 
+        helper :: Integer -> [Integer] -> [Integer] -> [Integer]
         helper ctr trav acc
-            | trav == [] = reverse acc
-            | ctr `mod` 2 == 1 = helper (ctr + 1) (tail trav) (head trav * 2 : acc)
+            | null trav = reverse acc
+            | odd ctr = helper (ctr + 1) (tail trav) (head trav * 2 : acc)
             | otherwise = helper (ctr + 1) (tail trav) (head trav : acc)
 
+
+doubleEveryOther' :: [Integer] -> [Integer]
+doubleEveryOther' = zipWith helper [1..] 
+    where
+        helper :: Integer -> Integer -> Integer
+        helper x y 
+            | even x = y*2
+            | otherwise = y
+
+-- Version using pattern matching within the helper
+doubleEveryOther'' :: [Integer] -> [Integer]
+doubleEveryOther'' lst = helper 0 lst [] 
+    where 
+        helper :: Integer -> [Integer] -> [Integer] -> [Integer]
+        helper _ [] acc = reverse acc
+        helper ctr (x:xs) acc 
+            | odd ctr = helper (ctr + 1) (xs) (x * 2 : acc)
+            | otherwise = helper (ctr + 1) (xs) (x:acc)
+
+-- Version using zipWith and cycle
+
+doubleEveryOther''' :: [Integer] -> [Integer]
+doubleEveryOther''' = zipWith helper (cycle [1,2])
```

Comment:
I see you're holding on for dear life on this idea of checking the parity of indices in your helper functions.
Just to finish this off here's what I actually had in mind:
```haskell
doubleEveryOther = zipWith (*) (cycle [1,2])
```

Reply:
Ah ok, that's what you meant. This is way way smarter. Did not see this one.